### ISSUE TITLE
[BUG/ISSUE] Stretched grid run errors with different regridding approaches

### ISSUE THREAD

**Issue by LukeFakes** [geoschem/gchp#273]
### What institution are you from?
University of York

### GEOS-Chem version
v14.0.1

### Description of the problem
I'm running a stretched grid simulation (c120, stretch 6) with standard emissions. The stretched run is using a c48 checkpoint file. 

I've tried two approaches for making stretch grid restarts, each getting the same error:
(1) gcpy.file_regrid (as in the readthedocs), which I'd successfully used for stretched grid runs in v13.4.0.
(2) the gridspec/sparselt/regrid_restart_file.py method as in [#161](https://github.com/geoschem/gcpy/issues/161). 

Less than a minute into the run, it fails with `Timer <GCHP> does not match start timer` error. 

```
Abort(0) on node 13 (rank 13 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 0) - process 13
pe=00014 FAIL at line=00167    BaseProfiler.F90                         <Timer <GCHP> does not match start timer <GCHPchem>>
pe=00014 FAIL at line=01814    MAPL_Generic.F90                         <status=1>
pe=00014 FAIL at line=00628    MAPL_CapGridComp.F90                     <status=1>
pe=00014 FAIL at line=00930    MAPL_CapGridComp.F90                     <status=1>
pe=00014 FAIL at line=00259    MAPL_Cap.F90                             <status=1>
pe=00014 FAIL at line=00218    MAPL_Cap.F90                             <status=1>
pe=00014 FAIL at line=00152    MAPL_Cap.F90                             <status=1>
pe=00014 FAIL at line=00129    MAPL_Cap.F90                             <status=1>
pe=00014 FAIL at line=00031    GCHPctm.F90                              <status=1>
```

Logfiles: [gcpy_regrid_logfiles.zip](https://github.com/geoschem/GCHP/files/10153937/gcpy_regrid.zip)
Restart: [Restart File](https://drive.google.com/file/d/1A8Y95PXerQfLJyxuMukey6vjvSbSDcMe/view?usp=share_link)

One additional question: 
The stretched grid tutorial on the GCHP readthedocs recommends using gcpy for restart regridding. Is this the preferred method or is it now gridspec/sparselt?

**Comment by Jourdan-He** [comment]
Hi @LukeFakes , the preferred way to regrid files is the second way you mentioned and the readthedocs will need to be updated, apologies for the confusion. The error for your simulation should be the first line of FAIL in the log file instead of the part you pasted. Could you share your GCHP configuration files (setCommonRunSettings.sh, cap_restart, and your run script)?

**Comment by LukeFakes** [comment]
Hi @Jourdan-He. Thanks, I'll use that method in the future. 
Sorry, I hadn't seen that it failed earlier in the log file with a "Factories not equal" error, which I haven't seen before.
Here are the config files:
[config_files.zip](https://github.com/geoschem/GCHP/files/10155711/config_files.zip)

**Comment by yantosca** [comment]
Thanks for writing @LukeFakes.  I've just pushed a fix to GCPy to the dev branch (commit 3b69d1670a98c6289b7dff6208ef1d8704a23e47) that includes a sample script `examples/regrid_restart_ll_to_cs.py`.  This is for regridding lat-lon to cube-sphere but you can modify that for your purposes.

**Comment by LukeFakes** [comment]
Hi @yantosca, I'll try modifying that script and regridding again. Is the error due to something in the restart?

**Comment by yantosca** [comment]
@LukeFakes: I don't know if your error is caused by the restart.  The problem with the script (from what was before) was that I had to redefine the coordinate arrays (lon, lat, lev) becuase for some reason the sparselt routine did not do it.

**Comment by lizziel** [comment]
"Factories" is referring to grid factories in MAPL. Most likely the restart file grid in the file does not match the run resolution.

**Comment by kilicomu** [comment]
Just to +1 this. We've tried regridding with the traditional GCPy method (which used to work fine!), with the method outlined by @Jourdan-He in https://github.com/geoschem/gcpy/issues/161, and with a modified version of the script linked by @yantosca. All result in the same `Factories not equal` error.

For clarity, we've tried regridding a spun-up restart and an out-of-the box restart, and what we're trying to achieve is a C48 restart file => C120 with a stretch.

Is there any useful troubleshooting we can do to help figure this out?

**Comment by lizziel** [comment]
I'd recommend digging into the `factors not equal `error in MAPL to find out what exactly is not equal. This should shed light on what is wrong with the file. Perhaps it is as simple as missing metadata that is now required in MAPL. 14.0 has a MAPL version update so this is totally possible.

**Comment by kilicomu** [comment]
@lizziel I've made some progress and am working on a fix.

**Comment by lizziel** [comment]
Okay, great. Keep us posted.

**Comment by kilicomu** [comment]
@lizziel I think I've got to the bottom of this.

The stretched grid attributes in regridded files (following the `gridspec` approach) need to be renamed to be read by MAPL. MAPL wants to find attributes named `STRETCH_FACTOR`, `TARGET_LAT`, and `TARGET_LON`, what we have are attributes named `stretch_factor`, `target_latitude`, and `target_longitude` (e.g. [here](https://github.com/geoschem/MAPL/blob/77fb1d4390b9f33e1aaad25d628a21968ce00c98/base/MAPL_CubedSphereGridFactory.F90#L325-L327)).

This doesn't solve the issue, though. The stretch factor is ending up as the default value of `MAPL_UNDEFINED_REAL` when it comes to comparing the physical parameters of the restart file grid to the physical parameters of the MAPL internal grid created from reading the model config files (comparison [here](https://github.com/geoschem/MAPL/blob/77fb1d4390b9f33e1aaad25d628a21968ce00c98/base/MAPL_CubedSphereGridFactory.F90#L817-L818), called from [here](https://github.com/geoschem/MAPL/blob/77fb1d4390b9f33e1aaad25d628a21968ce00c98/generic/MAPL_Generic.F90#L6068)).

It's dying [here](https://github.com/geoschem/MAPL/blob/77fb1d4390b9f33e1aaad25d628a21968ce00c98/base/MAPL_CubedSphereGridFactory.F90#L328-L353), where it tries to `type is (real(kind=REAL32))` the stretched grid attributes and only if they are `real(kind=REAL32)` will it set the stretched grid attributes in the factory. Changing these to `REAL64` resolves that problem, but leaves the outstanding question of whether or not these being 64 bit real attributes in the regridded restart is intentional.

To summarise:

1. Regridding process needs revisiting to ensure correct attribute names
2. Types of stretched grid attributes added by the regridding process need to be 32 bit reals in order to pass the MAPL `select type` on the stretched grid parameters, or the MAPL code needs to change to accept 64 bit reals

I'd favour changing the regridding process to match the MAPL code as is, and I'm happy to help with that if it's useful for me to do so.

**Comment by lizziel** [comment]
Thanks @kilicomu, this is great detective work. I agree it is fine to keep the parameters 32 bit to be compatible with MAPL. I don't see why we would need the lat, lon, or stretch parameters to be such high precision that they require 64 bit.

That would be great if you could help with updating this. @Jourdan-He was going to take a look but might be busy with other things.

**Comment by kilicomu** [comment]
No problem; I've put it on my list.

**Comment by mje1398** [comment]
I’m wondering whether we need to have a stretched grid run in the unit testing for gchp? This is the main use case for us (stretched grid) and it would have been nice if this problem had been caught at the unit testing stage rather than when somebody tried to do a relatively vanilla stretches grid run?

**Comment by yantosca** [comment]
Thanks @mje1398.  We can try to add a stretched grid integration test but it may take us a while given assigned priorities.

I have been working on updating the integration test scripts so maybe this is something I can look into as time allows.

**Comment by Jourdan-He** [comment]
@kilicomu Thanks so much for your detailed explanation! @lizziel I will work on fixing the attribute names and types.

**Comment by kilicomu** [comment]
@Jourdan-He I'm just taking a look at what I think is another problem with this process. I'll report back in an hour or so...

**Comment by kilicomu** [comment]
Okay, what I posted earlier wasn't quite correct!

I was testing with a regridded file that was created by GCPy. With GCPy regridded files, the issues above are correctly reported.

With the current `gridspec` approach, however, the issue is that the `STRETCH_FACTOR`, `TARGET_LAT`, and `TARGET_LON` attributes will only be added to the output regridded file if the attribute template file passed to `regrid_restart_file.py` contains those attributes, and the attribute values also need to be set correctly.

I `ncatted` the attributes into the output of the `gridspec` approach and the model runs fine without any changes.

So...

I propose that the right place to make the change to solve this issue is in GCPy. Either we want to go with the `gridspec` approach and we ought to integrate that approach into GCPy to make it more straightforward for users, or we keep the existing regridding routine and modify it to produce output that is accepted by MAPL.

I'm happy to work on either, but it would be good to figure out where you guys think the right place is for this to go and which method you'd want to work with.

**Comment by yantosca** [comment]
@kilicomu: I have created an example script for regridding lat-lon to CS with gridspec/sparselt.  You can use this as a starting point.  I think the gridspec/sparselt approach is probably the way to go in the future.
https://github.com/geoschem/gcpy/blob/dev/examples/working_with_files/regrid_restart_ll_to_cs.py

**Comment by yantosca** [comment]
@kilicomu: FYI, I wasn't aware of the updates in the standard "regrid_restart_file.py".  If you are working off that file that is OK too.

**Comment by lizziel** [comment]
My preference is to have the capability to regrid files built into GCPy. The documentation would then be within GCPy ReadTheDocs.

**Comment by kilicomu** [comment]
I've got a working fix to the `regrid_restart_file.py` example script - I'll tidy it up and try to PR it this week. After that maybe it's worth thinking about better integrating it with GCPy as the old method was. If I can make time I'll look into it.

**Comment by kilicomu** [comment]
PR to GCPy is [here](https://github.com/geoschem/gcpy/pull/190).

**Comment by kilicomu** [comment]
@lizziel @yantosca I think this needs to stay open until I've PRed some documentation for the modified file regridding process!

**Comment by lizziel** [comment]
Documentation is now included in GCPy ReadTheDocs.
